### PR TITLE
Integrate Web Worker Offloading with Google Site Kit

### LINKED ISSUE #1455: Offloading Google Analytics (gtag) to a Web Worker
## Feature Description

The Web Worker Offloading plugin is now merged into trunk (https://github.com/WordPress/performance/pull/1247). However, for it to be utilized a plugin author has to manually make the `web-worker-offloading` script a dependency of the script intended to be offloaded. As mentioned in https://github.com/WordPress/performance/pull/1247#issuecomment-2249065318, we should look for opportunities to offload existing scripts automatically as much as possible. From Partytown's [Common Services](https://partytown.builder.io/common-services) page, they list out 3rd party services known to be compatible. Included on that list is Google Tag Manager, which I've found to be responsible for INP issues on highly-trafficked sites. Partytown has [documentation](https://partytown.builder.io/google-tag-manager) for how to integrate with Google Tag Manager. 

The challenge for automatically implementing the Partytown integration with Google Tag Manager is that the plugins (and themes) which add gtag do so in inconsistent ways. The scripts could be registered properly via `wp_enqueue_script()` but the handles won't be consistent. Or a plugin may manually print the script at `wp_head`. 

In the case where a plugin properly enqueues the script, we can look at WPdirectory for the most common script handles used. Otherwise, for manually-printed scripts the alternative would be for Web Worker Offloading to integrate with Optimization Detective to register a tag visitor that looks for a `SCRIPT` tag with a `src` pointing to `https://www.googletagmanager.com/gtag/js`, and when present, inject the Partytown script (if not already present) and update the `SCRIPT` tag attributes as necessary.

In looking at [WPdirectory](https://wpdirectory.net/search/01J4VXSTK5GV5889HZXS4XD2CM) for plugins with at least 100K installs, the script handles used are:

Plugin | Handle | Install Count | Source
--|--|--:|--
[WooCommerce](https://wordpress.org/plugins/woocommerce/) ([Google Analytics for WooCommerce](https://wordpress.org/plugins/woocommerce-google-analytics-integration/)) | `google-tag-manager` | 5,000,000 (200,000) | [Trac](https://plugins.trac.wordpress.org/browser/woocommerce/tags/9.3.3/src/Blocks/Domain/Services/GoogleAnalytics.php#L50) ([Trac](https://plugins.trac.wordpress.org/browser/woocommerce-google-analytics-integration/tags/2.1.7/includes/class-wc-google-gtag-js.php#L66))
[Site Kit by Google](https://wordpress.org/plugins/google-site-kit/) | `google_gtagjs` | 4,000,000 | [Trac](https://plugins.trac.wordpress.org/browser/google-site-kit/tags/1.136.0/includes/Core/Tags/GTag.php#L111)
[Rank Math SEO](https://wordpress.org/plugins/seo-by-rank-math/) | `google_gtagjs` | 2,000,000 | [Trac](https://plugins.trac.wordpress.org/browser/seo-by-rank-math/tags/1.0.228/includes/modules/analytics/class-gtag.php#L180)

These are _not_ using GTM:

image

These add up to 11.2 million installs.

The following plugins print the script without using `WP_Scripts`:

Plugin | Install Count
--|--:
Jetpack _(legacy module)_ | 4,000,000
Complianz – GDPR/CCPA Cookie Consent | 900,000
Google Listings & Ads | 900,000
LightStart | 700,000
GA Google Analytics | 600,000
SEOPress | 300,000
Blocksy Companion | 200,000
WooCommerce Checkout & Funnel Builder by CartFlows | 200,000
Orbit Fox by ThemeIsle | 200,000
CTX Feed | 100,000
SEO Plugin by Squirrly SEO | 200,000
VK All in One Expansion Unit | 100,000

Excluding the Jetpack legacy module, these total up to 4.4 million installs.

So by starting out just targeting scripts registered via `WP_Scripts`, we can handle ~72% of the installs.

So the only script handles used are `google-tag-manager` and `google_gtagjs`. However, we could discover other script handles by looping over `wp_scripts()->registered` to find any other dependencies with a `src` beginning with `https://www.googletagmanager.com/gtag/js` at runtime. We can then add `web-worker-offloading` as a dependency for this registered script.

However, adding the script dependency is half of what we need to do. We also need to make sure that the inline script also gets the `text/partytown` type:

```html
  <script type="text/partytown">
    window.dataLayer = window.dataLayer || [];
    window.gtag = function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'YOUR-ID-HERE');
  </script>
```

WooCommerce adds this as an inline after script. As does Site Kit. It seems that "Google Analytics for WooCommerce" adds another script with its own inline script, so it may not be as straightforward to offload to a worker.

<details><summary><code>WC_Google_Gtag_JS::enquque_tracker()</code> [<em>sic</em>] method in <code>woocommerce-google-analytics-integration/includes/class-wc-google-gtag-js.php</code></summary>

```php
	public function enquque_tracker(): void {
		wp_enqueue_script(
			'google-tag-manager',
			'https://www.googletagmanager.com/gtag/js?id=' . self::get( 'ga_id' ),
			array(),
			null,
			array(
				'strategy' => 'async',
			)
		);
		// tracker.js needs to be executed ASAP, the remaining bits for main.js could be deffered,
		// but to reduce the traffic, we ship it all together.
		wp_enqueue_script(
			$this->script_handle,
			Plugin::get_instance()->get_js_asset_url( 'main.js' ),
			array(
				...Plugin::get_instance()->get_js_asset_dependencies( 'main' ),
				'google-tag-manager',
			),
			Plugin::get_instance()->get_js_asset_version( 'main' ),
			true
		);
		// Provide tracker's configuration.
		wp_add_inline_script(
			$this->script_handle,
			sprintf(
				'var wcgai = {config: %s};',
				wp_json_encode( $this->get_analytics_config() )
			),
			'before'
		);
	}
```

</details> 

But in the case of WooCommerce and Site Kit doing things in a more straightforward case: in order to add the `type="text/partytown"` to the inline script, we can do so via filtering `wp_inline_script_attributes`. For example:

```php
add_filter( 'wp_inline_script_attributes', static function ( $attributes, $data ) {
	if (
		array_key_exists( 'id', $attributes ) &&
		in_array( $attributes['id'], array( 'google-tag-manager-js-after', 'google_gtagjs-js-after' ), true ) &&
		str_contains( $data, 'dataLayer' )
	) {
		$attributes['type'] = 'text/partytown';
	}
	return $attributes;
} );
```

### PR DESCRIPTION
**Pull request by westonruter**
Fixes #1455

Depends on:

- [x] https://github.com/google/site-kit-wp/pull/9726

<details><summary>Diff in page source when activating</summary>

```diff
--- wwo-disabled.html	2024-11-20 18:13:10.158395531 -0800
+++ wwo-enabled.html	2024-11-20 18:12:37.771394543 -0800
@@ -12,7 +12,7 @@

 	<meta name='robots' content='index, follow, max-image-preview:large, max-snippet:-1, max-video-preview:-1' />
 	<!-- Google tag (gtag.js) Consent Mode dataLayer added by Site Kit -->
-	<script id="google_gtagjs-js-consent-mode-data-layer">
+	<script id="google_gtagjs-js-consent-mode-data-layer" type="text/partytown">
 		window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}
 		gtag('consent', 'default', {"ad_personalization":"denied","ad_storage":"denied","ad_user_data":"denied","analytics_storage":"denied","functionality_storage":"denied","security_storage":"denied","personalization_storage":"denied","region":["AT","BE","BG","CH","CY","CZ","DE","DK","EE","ES","FI","FR","GB","GR","HR","HU","IE","IS","IT","LI","LT","LU","LV","MT","NL","NO","PL","PT","RO","SE","SI","SK"],"wait_for_update":500});
 		window._googlesitekitConsentCategoryMap = {"statistics":["analytics_storage"],"marketing":["ad_storage","ad_user_data","ad_personalization"],"functional":["functionality_storage","security_storage"],"preferences":["personalization_storage"]};
@@ -198,11 +198,18 @@

 	</style>
 	<link rel='stylesheet' id='jetpack_css-css' href='https://example.com/wp-content/plugins/jetpack/css/jetpack.css?ver=13.8.1' media='all' />
+	<script id="web-worker-offloading-js-before">
+		window.partytown = {...(window.partytown || {}), ...{"lib":"\/wp-content\/plugins\/web-worker-offloading\/build\/","globalFns":["gtag"],"forward":["dataLayer.push"],"mainWindowAccessors":["_googlesitekitConsentCategoryMap","_googlesitekitConsents"]}};
+	</script>
+	<script id="web-worker-offloading-js-after">
+		/* Partytown 0.10.2-dev1727590485751 - MIT builder.io */
+		const t={preserveBehavior:!1},e=e=>{if("string"==typeof e)return[e,t];const[n,r=t]=e;return[n,{...t,...r}]},n=Object.freeze((()=>{const t=new Set;let e=[];do{Object.getOwnPropertyNames(e).forEach((n=>{"function"==typeof e[n]&&t.add(n)}))}while((e=Object.getPrototypeOf(e))!==Object.prototype);return Array.from(t)})());!function(t,r,o,i,a,s,c,d,l,p,u=t,f){function h(){f||(f=1,"/"==(c=(s.lib||"/~partytown/")+(s.debug?"debug/":""))[0]&&(l=r.querySelectorAll('script[type="text/partytown"]'),i!=t?i.dispatchEvent(new CustomEvent("pt1",{detail:t})):(d=setTimeout(v,999999999),r.addEventListener("pt0",w),a?y(1):o.serviceWorker?o.serviceWorker.register(c+(s.swPath||"partytown-sw.js"),{scope:c}).then((function(t){t.active?y():t.installing&&t.installing.addEventListener("statechange",(function(t){"activated"==t.target.state&&y()}))}),console.error):v())))}function y(e){p=r.createElement(e?"script":"iframe"),t._pttab=Date.now(),e||(p.style.display="block",p.style.width="0",p.style.height="0",p.style.border="0",p.style.visibility="hidden",p.setAttribute("aria-hidden",!0)),p.src=c+"partytown-"+(e?"atomics.js?v=0.10.2-dev1727590485751":"sandbox-sw.html?"+t._pttab),r.querySelector(s.sandboxParent||"body").appendChild(p)}function v(n,o){for(w(),i==t&&(s.forward||[]).map((function(n){const[r]=e(n);delete t[r.split(".")[0]]})),n=0;n<l.length;n++)(o=r.createElement("script")).innerHTML=l[n].innerHTML,o.nonce=s.nonce,r.head.appendChild(o);p&&p.parentNode.removeChild(p)}function w(){clearTimeout(d)}s=t.partytown||{},i==t&&(s.forward||[]).map((function(r){const[o,{preserveBehavior:i}]=e(r);u=t,o.split(".").map((function(e,r,o){var a;u=u[o[r]]=r+1<o.length?u[o[r]]||(a=o[r+1],n.includes(a)?[]:{}):(()=>{let e=null;if(i){const{methodOrProperty:n,thisObject:r}=((t,e)=>{let n=t;for(let t=0;t<e.length-1;t+=1)n=n[e[t]];return{thisObject:n,methodOrProperty:e.length>0?n[e[e.length-1]]:void 0}})(t,o);"function"==typeof n&&(e=(...t)=>n.apply(r,...t))}return function(){let n;return e&&(n=e(arguments)),(t._ptf=t._ptf||[]).push(o,arguments),n}})()}))})),"complete"==r.readyState?h():(t.addEventListener("DOMContentLoaded",h),t.addEventListener("load",h))}(window,document,navigator,top,window.crossOriginIsolated);
+	</script>
 	<script src="https://example.com/wp-content/themes/twentytwenty/assets/js/index.js?ver=2.7" id="twentytwenty-js-js" defer data-wp-strategy="defer"></script>

 	<!-- Google tag (gtag.js) snippet added by Site Kit -->
-	<script src="https://www.googletagmanager.com/gtag/js?id=GT-1234567" id="google_gtagjs-js" async></script>
-	<script id="google_gtagjs-js-after">
+	<script type="text/partytown" src="https://www.googletagmanager.com/gtag/js?id=GT-1234567" id="google_gtagjs-js" async></script>
+	<script id="google_gtagjs-js-after" type="text/partytown">
 		window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}
 		gtag("js", new Date());
 		gtag("set", "developer_id.dZTNiMT", true);
@@ -225,6 +232,7 @@
 	<meta name="apple-mobile-web-app-title" content="Weston Ruter">
 	<meta name="application-name" content="Weston Ruter">
 	<!-- Stream WordPress user activity plugin v4.0.2 -->
+	<meta name="generator" content="web-worker-offloading 0.1.1">
 	<meta name="generator" content="webp-uploads 2.2.0">
 	<script>document.documentElement.className = document.documentElement.className.replace( 'no-js', 'js' );</script>
 	<meta name="generator" content="image-prioritizer 0.1.4">
```

</details>

### CODE REVIEW

File: plugins/web-worker-offloading/third-party/google-site-kit.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	$configuration['mainWindowAccessors'][] = '_googlesitekitConsents';
+
+	return $configuration;
+}
+add_filter( 'plwwo_configuration', 'plwwo_google_site_kit_configure' );
+
+plwwo_mark_scripts_for_offloading(
+	array(
+		'google_gtagjs',
+		'googlesitekit-consent-mode',
```
Comment (westonruter) [inline plugins/web-worker-offloading/third-party/google-site-kit.php:49]: I don't know how the Consent Mode is supposed to work, especially as it relates to the WP Consent API plugin. On my test site I didn't get any consent banner even without WWO active, so I don't know what to look for to see if it is working as expected.

**Comment by westonruter** [comment]
@joemcgill @aaemnnosttv If someone from the Site Kit team could test that this is working as expected with a pre-release version of Site Kit, that would be very helpful.

**Comment by westonruter** [comment]
Build for testing: [web-worker-offloading.zip](https://github.com/user-attachments/files/17981166/web-worker-offloading.zip)

File: plugins/web-worker-offloading/third-party/google-site-kit.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * @link https://partytown.builder.io/google-tag-manager#forward-events
+ *
+ * @param array<string, mixed>|mixed $configuration Configuration.
+ * @return array<string, mixed> Configuration.
+ */
+function plwwo_google_site_kit_configure( $configuration ): array {
+	$configuration = (array) $configuration;
+
+	$configuration['globalFns'][] = 'gtag'; // Because gtag() is defined in one script and called in another.
+	$configuration['forward'][]   = 'dataLayer.push'; // Because the Partytown integration has this in its example config.
```
Comment (westonruter) [inline plugins/web-worker-offloading/third-party/google-site-kit.php]: Per https://github.com/WordPress/performance/pull/1685/files#r1884327683:

```suggestion
	$configuration['forward'][]   = 'dataLayer.push'; // See <https://partytown.builder.io/forwarding-event>.
```

File: plugins/web-worker-offloading/third-party/google-site-kit.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * @param array<string, mixed>|mixed $configuration Configuration.
+ * @return array<string, mixed> Configuration.
+ */
+function plwwo_google_site_kit_configure( $configuration ): array {
+	$configuration = (array) $configuration;
+
+	$configuration['globalFns'][] = 'gtag'; // Because gtag() is defined in one script and called in another.
+	$configuration['forward'][]   = 'dataLayer.push'; // See <https://partytown.builder.io/forwarding-event>.
+
+	// See <https://github.com/google/site-kit-wp/blob/440dcd6e8289a8776acc60ecb3c13f99bfd0ac5a/includes/Core/Consent_Mode/Consent_Mode.php#L239-L240>.
```
Comment (westonruter) [inline plugins/web-worker-offloading/third-party/google-site-kit.php]:
```suggestion
	// See <https://github.com/google/site-kit-wp/blob/abbb74ff21f98a8779fbab0eeb9a16279a122bc4/includes/Core/Consent_Mode/Consent_Mode.php#L244-L259>.
```

**Comment by westonruter** [comment]
Updated build for testing which includes additional compatibility with the WP Consent API: [web-worker-offloading.zip](https://github.com/user-attachments/files/18130845/web-worker-offloading.zip)

File: plugins/web-worker-offloading/third-party/google-site-kit.php (hunk 1 above)
Comment (adamsilverstein) [reply plugins/web-worker-offloading/third-party/google-site-kit.php:49]: you may need to have this plugin installed? https://wordpress.org/plugins/wp-consent-api/

File: plugins/web-worker-offloading/third-party/google-site-kit.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+plwwo_mark_scripts_for_offloading(
+	array(
+		'google_gtagjs',
+		'googlesitekit-consent-mode',
+	)
+);
+
+/**
+ * Filters inline script attributes to offload Rank Math's GTag script tag to Partytown.
```
Comment (adamsilverstein) [inline plugins/web-worker-offloading/third-party/google-site-kit.php]:
```suggestion
 * Filters inline script attributes to offload Google Site Kit's GTag script tag to Partytown.
```

File: plugins/web-worker-offloading/third-party/google-site-kit.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * @since n.e.x.t
+ * @access private
+ * @link https://github.com/rankmath/seo-by-rank-math/blob/c78adba6f78079f27ff1430fabb75c6ac3916240/includes/modules/analytics/class-gtag.php#L169-L174
```
Comment (adamsilverstein) [inline plugins/web-worker-offloading/third-party/google-site-kit.php]: this link is incorrect, remove?

**Review by adamsilverstein — COMMENTED** [review]
Looks good, left some small feedback.

File: plugins/web-worker-offloading/third-party/google-site-kit.php (hunk 1 above)
Comment (adamsilverstein) [reply plugins/web-worker-offloading/third-party/google-site-kit.php:49]: cc: @aaemnnosttv ^^

**Comment by aaemnnosttv** [comment]
@westonruter I gave this a few tests using the latest test build (https://github.com/WordPress/performance/pull/1686#issuecomment-2542095149) as well and it seemed to work as expected. However, on closer inspection, it seems there may be an issue with inline scripts as the inline script that accompanies the Google tag doesn't seem to be running (e.g. `window.gtag` is undefined).

image

File: plugins/web-worker-offloading/third-party/google-site-kit.php (hunk 5 above)
Comment (westonruter) [reply plugins/web-worker-offloading/third-party/google-site-kit.php]: Oops. This should be it:

```suggestion
 * @link https://github.com/google/site-kit-wp/blob/abbb74ff21f98a8779fbab0eeb9a16279a122bc4/includes/Core/Consent_Mode/Consent_Mode.php#L244-L259
```

**Comment by westonruter** [comment]
@aaemnnosttv Is this anywhere I can test?

**Comment by westonruter** [comment]
Another build for testing: [web-worker-offloading.zip](https://github.com/user-attachments/files/18132524/web-worker-offloading.zip)

File: plugins/web-worker-offloading/third-party/seo-by-rank-math.php [hunk 6]
```diff
@@ -24,7 +24,10 @@ function plwwo_rank_math_configure( $configuration ): array {
 	$configuration = (array) $configuration;
 
 	$configuration['globalFns'][] = 'gtag'; // Because gtag() is defined in one script and called in another.
-	$configuration['forward'][]   = 'dataLayer.push'; // See <https://partytown.builder.io/forwarding-event>.
+
+	// Expose on the main tread. See <https://partytown.builder.io/forwarding-event>.
+	$configuration['forward'][] = 'dataLayer.push';
+	$configuration['forward'][] = 'gtag';
```
Comment (westonruter) [inline plugins/web-worker-offloading/third-party/seo-by-rank-math.php:30]: The addition here means that scripts in the main thread will now be able to call `gtag()` whereas previously it was only available to other scripts offloaded to a worker.

File: plugins/web-worker-offloading/third-party/woocommerce.php [hunk 7]
```diff
@@ -23,10 +23,15 @@
 function plwwo_woocommerce_configure( $configuration ): array {
 	$configuration = (array) $configuration;
 
+	$configuration['globalFns'][] = 'gtag'; // Allow calling from other Partytown scripts.
+
+	// Expose on the main tread. See <https://partytown.builder.io/forwarding-event>.
+	$configuration['forward'][] = 'dataLayer.push';
+	$configuration['forward'][] = 'gtag';
```
Comment (westonruter) [inline plugins/web-worker-offloading/third-party/woocommerce.php:30]: The addition of this line here means that scripts in the main thread will now be able to call `gtag()` whereas previously it was only available to other scripts offloaded to a worker.

Comment (aaemnnosttv) [reply plugins/web-worker-offloading/third-party/woocommerce.php:30]: I can confirm this works now 👍

**Comment by aaemnnosttv** [comment]
With the recent updates here and my understanding of the expectations, this seems to be working as expected 👍

**Review by adamsilverstein — APPROVED** [review]
Tremendous 🎉

**Pull request merged**